File: PrimeLoanPH/PBModules/PBGoods/PPGoodsDetailViewController.m
//
//  PPGoodsDetailViewController.m
//  PrimeLoanPH
//
//  Created by MacPing on 2023/10/31.
//

#import "PPGoodsDetailViewController.h"
#import "PPGoodsDetailHeader.h"
#import "PPGoodsDetailTableViewCell.h"
#import "PPDetailModel.h"
#import "PB_FunSureAlertView.h"
#import "PPBaseModel.h"

@interface PPGoodsDetailViewController ()

@property (nonatomic, strong) PPDetailModel *dataModel;
@property (nonatomic, strong) NSArray <PPDetailValuingModel *>*dataArray;
@property (nonatomic, copy) NSString *pb_t_de_nextStr;//下一步操作
@property (nonatomic, copy) NSString *pb_t_de_pouductOrderNo;//订单号
@property (nonatomic, assign) QMUIButton *pb_t_de_submitBtn;
@property (nonatomic, assign) QMUIButton *agreeBtn;
@property (nonatomic, assign) QMUIButton *pb_t_de_pAgreebutton;
@property (nonatomic, strong) PPGoodsDetailHeader *pb_t_de_headerView;

@property (nonatomic, assign) BOOL isFirst;

@end

@implementation PPGoodsDetailViewController


-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if(_isFirst == YES){
        [self requestMethod];
    }
    _isFirst = YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self ppInit];
    self.navBgColr = UIColor.clearColor;
    [self setShowBackBtn:YES];
    [self setNavTitle:@"Pocket Cash"];
    
}


- (void)ppInit {

    
    self.view.backgroundColor = PB_WhiteColor;
    self.pb_t_de_nextStr = @"";
    self.pb_t_de_pouductOrderNo = @"";
    self.pb_t_de_headerView = [[PPGoodsDetailHeader alloc] init];
    self.tableView.backgroundColor = PB_WhiteColor;
    self.tableView.tableHeaderView = self.pb_t_de_headerView;
    [self.tableView registerClass:[PPGoodsDetailTableViewCell class] forCellReuseIdentifier:PPGoodsDetailTableViewCellKey];
    [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(UIEdgeInsetsMake(- StatusBarHeightConstant, 0, PB_Ratio(100), 0));
    }];
    


    QMUIButton *submitButton = [QMUIButton buttonWithType:UIButtonTypeCustom];
    submitButton.backgroundColor = PP_AppColor;
    submitButton.layer.cornerRadius = PB_Ratio(22);
    submitButton.layer.masksToBounds = YES;
    [submitButton setTitle:@"Apply Now" forState:UIControlStateNormal];
    submitButton.titleLabel.font = UIFontMediumMake(PB_Ratio(18));
    [submitButton setTitleColor:PB_WhiteColor forState:UIControlStateNormal];
    [submitButton addTarget:self action:@selector(submitButtonSender) forControlEvents:UIControlEventTouchUpInside];
    
    
    
    //    
    
    QMUIButton *agreeButton = [QMUIButton buttonWithType:UIButtonTypeCustom];
    agreeButton.backgroundColor = UIColor.clearColor;

    [agreeButton setImage:[UIImage imageNamed:@"signin_icon_select"] forState:UIControlStateNormal];
    [agreeButton setImage:[UIImage imageNamed:@"signin_icon_select"] forState:UIControlStateHighlighted];
    [agreeButton addTarget:self action:@selector(agreeButtonSender:) forControlEvents:UIControlEventTouchUpInside];

    
    [agreeButton setImage:UIImageMake(@"signin_icon_selected") forState:UIControlStateSelected];
    agreeButton.selected = YES;
    //
    NSString *agree = @"I have read and agree to the Loan Agreement ";
    NSString *agree1 = @"Loan Agreement ";
    
    QMUIButton *pb_t_de_pAgreebutton = [QMUIButton buttonWithType:UIButtonTypeCustom];
    pb_t_de_pAgreebutton.backgroundColor = UIColor.clearColor;
    [pb_t_de_pAgreebutton setTitle:agree forState:UIControlStateNormal];
    pb_t_de_pAgreebutton.titleLabel.font = UIFontMake(PB_Ratio(12));
    [pb_t_de_pAgreebutton setTitleColor:PB_Gray_1_Color forState:UIControlStateNormal];
    [pb_t_de_pAgreebutton addTarget:self action:@selector(agreeButtonAction:) forControlEvents:UIControlEventTouchUpInside];
    
    
    
    
    NSAttributedString *attri = [PPTools pb_t_attriStringWithHexString:agree1 totalStr:agree norColor:PB_Gray_1_Color attriColor:PP_AppColor norFont:UIFontMake(PB_Ratio(12)) attriFont:UIFontMake(PB_Ratio(12)) underline:YES];
    [pb_t_de_pAgreebutton setAttributedTitle:attri forState:UIControlStateNormal];
    
    

    [self.view addSubview:submitButton];
    [self.view addSubview:agreeButton];
    [self.view addSubview:pb_t_de_pAgreebutton];
    self.pb_t_de_submitBtn = submitButton;
    self.agreeBtn = agreeButton;
    self.pb_t_de_pAgreebutton = pb_t_de_pAgreebutton;
    self.agreeBtn.hidden = YES;
    self.pb_t_de_pAgreebutton.hidden = YES;
    [submitButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(0);
        make.bottom.mas_offset(-PB_Ratio(50));
        make.width.mas_equalTo(PB_SW - PB_Ratio(47)*2);
        make.height.mas_equalTo(PB_Ratio(44));
    }];
    [agreeButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.height.mas_equalTo(PB_Ratio(21));
        make.left.mas_equalTo(PB_Ratio(31));
        make.top.mas_equalTo(submitButton.mas_bottom).offset(PB_Ratio(6));
    }];
    
    [pb_t_de_pAgreebutton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(agreeButton);
        make.left.mas_equalTo(agreeButton.mas_right);
        make.width.mas_lessThanOrEqualTo(PB_SW - PB_Ratio(67));
    }];
    [self requestMethod];
    [self setPpShowTableViewHeaderRefresh:YES];
}


- (void)requestMethod{

    if(_isFirst == NO){
        [QMUITips showLoading:PBLoading_TipMsg inView:self.view];
    }
    NSDictionary *p = @{
        @"foundation":PBStrFormat(self.goodsId)
    };
    [[PB_RequestHelper pb_instance] pb_postRequestWithUrlStr:PBURL_productDetailInfoUrl params:p commplete:^(NSDictionary * _Nullable result, NSInteger statusCode) {
        [QMUITips hideAllTips];
        if(result != nil){
            self.dataModel = [PPDetailModel yy_modelWithJSON:result];
            [self.pb_t_de_headerView pp_configData:self.dataModel];
            self.dataArray = @[];
            if(self.dataModel && self.dataModel.theoretical.valuing.count > 0){
                self.dataArray = self.dataModel.theoretical.valuing;
            }
            if(self.dataModel.theoretical.grant != nil){
                self.pb_t_de_nextStr = PBStrFormat(self.dataModel.theoretical.grant.availability);
            }else{
                self.pb_t_de_nextStr = @"";
            }
            if(![NSString PB_CheckStringIsEmpty:self.dataModel.theoretical.addressed.courses]){
                self.navTitle = self.dataModel.theoretical.addressed.courses;
            }
            self.pb_t_de_pouductOrderNo = PBStrFormat(self.dataModel.theoretical.addressed.enough);
            self.pb_t_de_submitBtn.hidden = self.dataArray.count > 0 ? NO : YES;
            if(![NSString PB_CheckStringIsEmpty:self.dataModel.theoretical.addressed.lobbying]){
                [self.pb_t_de_submitBtn setTitle:self.dataModel.theoretical.addressed.lobbying forState:UIControlStateNormal];
            }
            if([NSString PB_CheckStringIsEmpty:self.dataModel.theoretical.ethnic.age]){
                self.agreeBtn.hidden = YES;
                self.pb_t_de_pAgreebutton.hidden = YES;
            }else{
                self.agreeBtn.hidden = NO;
                self.pb_t_de_pAgreebutton.hidden = NO;
            }
        }
        [self.tableView reloadData];
        [self ppTableViewEndAllRefresh];
    } failure:^(NSError * _Nonnull error, NSInteger errorCode, NSString * _Nonnull errorStr) {
        [QMUITips showError:errorStr inView:self.view];
        [self ppTableViewEndAllRefresh];
    }];
}

- (void)requestToBorrow{
    [QMUITips showLoading:PBLoading_TipMsg inView:self.view];
//    NSDictionary *params = @{
//        @"medicinal":orderNo,//订单号
//        @"develop":PHStringForm(self.dataModel.should.level.develop),//金额
//        @"reveals":PHStringForm(self.dataModel.should.level.reveals),//借款期限
//        @"unique":@(self.dataModel.should.level.unique),//期限类型
//    };
    NSString *goodsOrderId = PBStrFormat(self.dataModel.theoretical.addressed.enough);
    NSString *goodsAmount = PBStrFormat(self.dataModel.theoretical.addressed.issue);
    NSString *goodsTerm = PBStrFormat(self.dataModel.theoretical.addressed.narratives);
    NSString *l_oanType = PBStrFormat(self.dataModel.theoretical.addressed.drury);

    NSDictionary *params = @{
        @"aged":goodsOrderId,//订单号
        @"issue":goodsAmount,//金额
        @"narratives":goodsTerm,//借款期限
        @"drury":l_oanType,//借款类型
    };

    [[PB_RequestHelper pb_instance] pb_postRequestWithUrlStr:PBURL_myOrderListItemLinkUrl params:params commplete:^(NSDictionary * _Nullable result, NSInteger statusCode) {
        [QMUITips hideAllTips];
        if(result != nil){
            PPBaseModel *model = [PPBaseModel yy_modelWithJSON:result];
            NSString *linkStr = PBStrFormat(model.theoretical.translated);
            [PB_APP_Control pb_t_goToModuleWithJudgeTypeStr:linkStr fromVC:self];
            [self pb_t_toRePortRiskDataToServeFromStep];
            [PB_GetVC pb_to_removeViewController:self];
        }
    } failure:^(NSError * _Nonnull error, NSInteger errorCode, NSString * _Nonnull errorStr) {
        [QMUITips showError:errorStr];
    }];
}


- (void)pb_t_de_TableViewHeaderRefreshMethod {
    [self requestMethod];
}


#pragma mark - tableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PPGoodsDetailTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:PPGoodsDetailTableViewCellKey forIndexPath:indexPath];
    [cell pb_configWithCellData:self.dataArray[indexPath.row]];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return PB_Ratio(70);
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if(self.dataArray.count == 0)
        return;

    /**
     1、未完成--可点击
     2、已完成--可查看结果
     3、认证需要按照顺序
     */
    PPDetailValuingModel *model = self.dataArray[indexPath.row];
//    if([model.availability isEqualToString:@"ste"]){//第五步，绑卡H5
//        if([NSString PB_CheckStringIsEmpty:model.translated]){
//            [self submitButtonSender];
//        }else{
//            [PB_APP_Control pb_t_goToModuleWithJudgeTypeStr:model.translated fromVC:self];
//        }
//
//    }else{
//        if(model.acknowledges != 1){ //未完成
//            [self submitButtonSender];
//        }else{ //查看
//            NSString *currentStep = PBStrFormat(model.availability);
//            [self toAuthVCWithSetpId:currentStep];
//        }
//    }
    
    if(model.acknowledges != 1){ //未完成
        [self submitButtonSender];
    }else{ //查看
        if([model.availability isEqualToString:@"ste"]){//第五步，绑卡H5
            if([NSString PB_CheckStringIsEmpty:model.translated]){
                //详情查看链接为空，那么直接进行认证H5
                [self submitButtonSender];
            }else{
                //已经从H5认证过，那么进行查看H5详情
                [PB_APP_Control pb_t_goToModuleWithJudgeTypeStr:model.translated fromVC:self];
            }
        }else{
            NSString *currentStep = PBStrFormat(model.availability);
            [self toAuthVCWithSetpId:currentStep];
        }
    }

}

//进入认证页面的唯一入口
- (void)toAuthVCWithSetpId:(NSString *)stepId {
    NSString *productId = PBStrFormat(self.dataModel.theoretical.addressed.pivotal);
    NSString *orderNo = PBStrFormat(self.dataModel.theoretical.addressed.enough);
    //获取当前步骤属于第几步
    [PB_APP_Control pb_t_toCertifyStepIndexWithProductId:productId oId:orderNo stepStr:stepId fromVC:self];
}


#pragma mark - click

- (void)agreeButtonAction:(QMUIButton *)button {
    NSString *agreeUrlString = PBStrFormat(self.dataModel.theoretical.ethnic.funds);
    [PB_APP_Control pb_t_goToModuleWithJudgeTypeStr:agreeUrlString fromVC:self];
}

- (void)submitButtonSender {
    if(self.dataModel == nil){
        return;
    }
    if(self.dataModel.theoretical.valuing && self.dataModel.theoretical.valuing.count == 0){
        NSLog(@"认证项为空，不能 take order");
        return;
    }else{
        BOOL hasAllComplete = YES;
        //判断是否进行下单
        for (NSInteger i = 0; i < self.dataModel.theoretical.valuing.count; i++) {
            PPDetailValuingModel *model = self.dataModel.theoretical.valuing[i];
            if(model.acknowledges == 0){
                hasAllComplete = NO;
                break;
            }
        }
        if(hasAllComplete == YES){
            
            //下单操作
            QMUIModalPresentationViewController *qmAlert = [[QMUIModalPresentationViewController alloc] init];
            qmAlert.animationStyle = QMUIModalPresentationAnimationStylePopup;

            PB_FunSureAlertView *tipView = [[PB_FunSureAlertView alloc] init];
            qmAlert.contentView = tipView;
            //关闭点击背景自动隐藏
            qmAlert.modal = YES;
            [qmAlert showWithAnimated:YES completion:nil];
            PMMyWeekSelf
            tipView.myBlock = ^(NSInteger buttonIndex) {
                [qmAlert hideWithAnimated:YES completion:nil];
                if(buttonIndex == 50){
                    [weakSelf requestToBorrow];
                }
            };
            return;
        }
    }
    
    if([NSString PB_CheckStringIsEmpty:self.pb_t_de_nextStr]){
        NSLog(@"next auth is empty");
    }else{
        if([self.pb_t_de_nextStr isEqualToString:@"ste"]){ //h5
            if([NSString PB_CheckStringIsEmpty:self.dataModel.theoretical.grant.translated] == NO){
                NSString *url = PBStrFormat(self.dataModel.theoretical.grant.translated);
                [PB_APP_Control pb_t_goToModuleWithJudgeTypeStr:url fromVC:self];
            }
        }else{
           [self toAuthVCWithSetpId:self.pb_t_de_nextStr];
        }
    }
}

- (void)agreeButtonSender:(QMUIButton *)button {
    button.selected = !button.selected;
    self.pb_t_de_submitBtn.enabled = button.selected;
}



//#pragma mark - risk
- (void)pb_t_toRePortRiskDataToServeFromStep {
    NSString *goodsOrderId = PBStrFormat(self.dataModel.theoretical.addressed.enough);
    NSDictionary *riskDict = @{
        @"speak":PBStrFormat([PB_timeHelper pb_t_getCurrentStampTimeString]),
        @"advantage":PBStrFormat([PB_timeHelper pb_t_getCurrentStampTimeString]),
        @"rejection":@"9",
        @"constraining":PBStrFormat(goodsOrderId)
    };
    [[PB_APP_Control instanceOnly] pb_t_toRePortRiskDataToServe:riskDict];
}

@end

File: PrimeLoanPH/PBModules/PBGoods/PPGoodsDetailViewController.swift
//
//  PPGoodsDetailViewController.swift
//  PrimeLoanPH
//

import SwiftUI
import UIKit

final class GoodsDetailViewState: ObservableObject {
    @Published var productTitle: String = ""
    /// 原固定文案「Maximum loan amount」，对应接口 `addressed.aiming`
    @Published var amountCaptionText: String = ""
    @Published var amountText: String = ""
    /// 左列标题：`rich.can.age`；数值：`rich.can.view`
    @Published var loanTermTitle: String = ""
    @Published var loanTermValue: String = ""
    /// 右列标题：`rich.subtractive.age`；数值：`rich.subtractive.view`
    @Published var interestRateTitle: String = ""
    @Published var interestRateValue: String = ""
    @Published var logoURLs: [String] = []
    @Published var steps: [PPDetailValuingModel] = []
    @Published var showAgreement: Bool = false
    @Published var agreementChecked: Bool = true
    @Published var applyTitle: String = ""

    var applyEnabled: Bool {
        !showAgreement || agreementChecked
    }
}

@objc(PPGoodsDetailViewController)
final class PPGoodsDetailViewController: PPBaseViewController {

    @objc var goodsId: String = ""

    private let state = GoodsDetailViewState()
    private var dataModel: PPDetailModel?
    private var nextStep: String = ""
    private var orderNo: String = ""
    private var didAppearOnce = false
    private var hostingController: UIHostingController<GoodsDetailRootView>?

    @objc(initWithPBTableViewOfGroupStyle:)
    convenience init(pbTableViewOfGroupStyle: Bool) {
        self.init(nibName: nil, bundle: nil)
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        
        showNavBar = true
        showBackBtn = true
        navTitle = ""

        let pageBg = UIColor.pbColorBackHexStr("#FBF6E7")
        view.backgroundColor = pageBg
        navigationController?.view.backgroundColor = pageBg

        let root = GoodsDetailRootView(
            state: state,
            onStepTap: { [weak self] idx in self?.handleStepTap(index: idx) },
            onAgreementTap: { [weak self] in self?.openAgreement() },
            onAgreementToggle: { [weak self] in
                guard let self else { return }
                self.state.agreementChecked.toggle()
            },
            onApplyTap: { [weak self] in self?.handleApply() }
        )

        let host = UIHostingController(rootView: root)
        host.view.backgroundColor = .clear
        host.view.clipsToBounds = true
        hostingController = host
        addChild(host)
        host.view.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(host.view)
        NSLayoutConstraint.activate([
            host.view.topAnchor.constraint(equalTo: pb_navigationBarContainerView.bottomAnchor),
            host.view.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            host.view.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            host.view.trailingAnchor.constraint(equalTo: view.trailingAnchor)
        ])
        host.didMove(toParent: self)
        pp_bringCustomNavigationBarToFront()

        requestDetail(showLoading: false)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if didAppearOnce {
            requestDetail(showLoading: true)
        }
        didAppearOnce = true
    }

    private func requestDetail(showLoading: Bool) {
        if showLoading, let content = hostingController?.view {
            PB_NativeTipsHelper.pb_showLoading(in: content)
        }

        let params: [String: Any] = [
            "foundation": goodsId
        ]
        let url = PB_API_ProductDetailInfoURL() as String
        PB_RequestHelper.pb_instance().pb_postRequest(withUrlStr: url, params: params, commplete: { [weak self] result, _ in
            DispatchQueue.main.async {
                PB_NativeTipsHelper.pb_hideAllLoading()
                self?.pp_bringCustomNavigationBarToFront()
                guard let self else { return }
                guard let result else { return }
                if let model = PPDetailModel.yy_model(withJSON: result) {
                    self.applyDetailModel(model)
                }
            }
        }, failure: { [weak self] _, _, msg in
            DispatchQueue.main.async {
                PB_NativeTipsHelper.pb_hideAllLoading()
                self?.pp_bringCustomNavigationBarToFront()
                let text = (msg as String?) ?? ""
                if !text.isEmpty {
                    PB_NativeTipsHelper.pb_presentAlert(withMessage: text)
                }
            }
        })
    }

    private func applyDetailModel(_ model: PPDetailModel) {
        dataModel = model
        let addressed = model.theoretical.addressed
        let rich = addressed?.rich

        let coursesTitle = (addressed?.courses ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.productTitle = coursesTitle
        navTitle = coursesTitle

        let aiming = (addressed?.aiming ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.amountCaptionText = aiming
        let unit = (addressed?.trend ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let issue = (addressed?.issue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.amountText = unit + issue

        state.loanTermTitle = (rich?.can?.age ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.loanTermValue = (rich?.can?.view ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.interestRateTitle = (rich?.subtractive?.age ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        state.interestRateValue = (rich?.subtractive?.view ?? "").trimmingCharacters(in: .whitespacesAndNewlines)

        state.steps = (model.theoretical.valuing as? [PPDetailValuingModel]) ?? []
        state.applyTitle = (addressed?.lobbying?.isEmpty == false) ? (addressed?.lobbying ?? "") : ""

        let agreementURL = model.theoretical.ethnic?.funds ?? ""
        state.showAgreement = !agreementURL.isEmpty
        state.agreementChecked = true

        if let logos = addressed?.currently as? [String] {
            state.logoURLs = logos.filter { !$0.isEmpty }
        } else {
            state.logoURLs = []
        }

        nextStep = model.theoretical.grant?.availability ?? ""
        orderNo = addressed?.enough ?? ""
    }

    private func openAgreement() {
        guard let url = dataModel?.theoretical.ethnic?.funds, !url.isEmpty else { return }
        PB_APP_Control.pb_t_goToModule(withJudgeTypeStr: url, fromVC: self)
    }

    private func handleStepTap(index: Int) {
        guard index >= 0 && index < state.steps.count else { return }
        let model = state.steps[index]
        if model.acknowledges != 1 {
            handleApply()
            return
        }

        if model.availability == "ste" {
            let link = model.translated ?? ""
            if link.isEmpty {
                handleApply()
            } else {
                PB_APP_Control.pb_t_goToModule(withJudgeTypeStr: link, fromVC: self)
            }
            return
        }

        toAuthVC(stepId: model.availability ?? "")
    }

    private func toAuthVC(stepId: String) {
        guard !stepId.isEmpty else { return }
        let productId = dataModel?.theoretical.addressed?.pivotal ?? ""
        let oId = dataModel?.theoretical.addressed?.enough ?? ""
        PB_APP_Control.pb_t_toCertifyStepIndex(withProductId: productId, oId: oId, stepStr: stepId, fromVC: self)
    }

    private func handleApply() {
        guard state.applyEnabled else { return }
        guard let model = dataModel else { return }

        let steps = (model.theoretical.valuing as? [PPDetailValuingModel]) ?? []
        guard !steps.isEmpty else { return }

        let hasAllComplete = !steps.contains { $0.acknowledges == 0 }
        if hasAllComplete {
//            let alert = UIAlertController(title: "Notice", message: "Are you sure you want to apply now?", preferredStyle: .alert)
//            alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
//            alert.addAction(UIAlertAction(title: "Apply", style: .default, handler: { [weak self] _ in
//                self?.requestToBorrow()
//            }))
//            present(alert, animated: true)
            
            requestToBorrow()
            return
        }

        guard !nextStep.isEmpty else { return }
        if nextStep == "ste" {
            let link = dataModel?.theoretical.grant?.translated ?? ""
            if !link.isEmpty {
                PB_APP_Control.pb_t_goToModule(withJudgeTypeStr: link, fromVC: self)
            }
        } else {
            toAuthVC(stepId: nextStep)
        }
    }

    private func requestToBorrow() {
        if let content = hostingController?.view {
            PB_NativeTipsHelper.pb_showLoading(in: content)
        }

        let params: [String: Any] = [
            "aged": dataModel?.theoretical.addressed?.enough ?? "",
            "issue": dataModel?.theoretical.addressed?.issue ?? "",
            "narratives": dataModel?.theoretical.addressed?.narratives ?? "",
            "drury": dataModel?.theoretical.addressed?.drury ?? ""
        ]

        let url = PB_API_MyOrderListItemLinkURL() as String
        PB_RequestHelper.pb_instance().pb_postRequest(withUrlStr: url, params: params, commplete: { [weak self] result, _ in
            DispatchQueue.main.async {
                PB_NativeTipsHelper.pb_hideAllLoading()
                self?.pp_bringCustomNavigationBarToFront()
                guard let self else { return }
                if let theoretical = result?["theoretical"] as? [String: Any],
                   let link = theoretical["translated"] as? String,
                   !link.isEmpty {
                    self.reportRiskFromStep()
                    
                    self.navigationController?.popViewController(animated: false)

                    if let tvc = PB_GetVC.pb_to_getCurrentViewController() as? PPBaseViewController {
                        PB_APP_Control.pb_t_goToModule(withJudgeTypeStr: link, fromVC: tvc)

                    }

                }
            }
        }, failure: { [weak self] _, _, msg in
            DispatchQueue.main.async {
                PB_NativeTipsHelper.pb_hideAllLoading()
                self?.pp_bringCustomNavigationBarToFront()
                let text = (msg as String?) ?? ""
                if !text.isEmpty {
                    PB_NativeTipsHelper.pb_presentAlert(withMessage: text)
                }
            }
        })
    }

    private func reportRiskFromStep() {
        let order = dataModel?.theoretical.addressed?.enough ?? orderNo
        let stamp = PB_timeHelper.pb_t_getCurrentStampTimeString()
        let risk: [String: Any] = [
            "speak": stamp,
            "advantage": stamp,
            "rejection": "9",
            "constraining": order
        ]
        PB_APP_Control.instanceOnly().pb_t_toRePortRiskDataToServe(risk)
    }
}
